kong/plugins/moesif/moesif_gov.lua

```diff
@@ -0,0 +1,310 @@
+        end
+    else
+        if conf.debug then
+            ngx_log(ngx.DEBUG, "[moesif] Successfully send request to fetch the governance rules " , ok)
+        end
+    end
+
+    -- Read the response
+    local governance_rules_response = helper.read_socket_data(sock)
+
+    ok, err = sock:setkeepalive(conf.keepalive)
+    if not ok then
+        -- Set fetch_governance_rule to true, to be able to fetch governance rule next time
+        fetch_governance_rules[hash_key] = true
+        if conf.debug then
+            ngx_log(ngx_log_ERR, "[moesif] failed to keepalive to " .. parsed_url.host .. ":" .. tostring(parsed_url.port) .. ": ", err)
+        end
+
+        local close_ok, close_err = sock:close()
+        if not close_ok then
+            if conf.debug then
+                ngx_log(ngx_log_ERR,"[moesif] Failed to manually close socket connection ", close_err)
+            end
+        else
+            if conf.debug then
+                ngx_log(ngx.DEBUG,"[moesif] success closing socket connection manually ")
+            end
+        end
+    else
+        if conf.debug then
+            ngx_log(ngx.DEBUG,"[moesif] success keep-alive", ok)
+        end
+    end
+
+    -- Get the governance rules
+    local governance_rules = {}
+    local response_body = cjson.decode(governance_rules_response:match("(%[.*])"))
+    for k, rule in pairs(response_body) do
+        governance_rules[rule["_id"]] = rule
+    end
+
+    -- Save the governance rule in the dictionary
+    if type(governance_rules) == "table" and next(governance_rules) ~= nil then
+        governance_rules_hashes[hash_key] = governance_rules
+    end
+
+    -- Read the Response tag
+    local rules_etag = string.match(governance_rules_response, "Tag%s*:%s*(.-)\n")
+    governance_rules_etags[hash_key] = rules_etag
+
+    -- Get app config
+    log.get_config_for_rules(conf)
+
+    -- Set fetch_governance_rule to true, to be able to fetch governance rule next time
+    fetch_governance_rules[hash_key] = true
+end
+
+
+-- Fetch governance rule of the entity for which block is true
+-- @param `entity_rules`      List of entity rules
+-- @param `governance_rules`  List of governance rules
+function fetch_entity_governance_rule(entity_rules, governance_rules)
+    for _, rule in pairs(entity_rules) do
+        local rule_id = rule["rules"]
+        if rule_id ~= nil and governance_rules[rule_id] ~= nil then 
+            if governance_rules[rule_id]["block"] then 
+                return rule
+            end
+        end
+    end
+end
+
+-- Replace body value in the response body for the short-circuited request
+-- @param `body_table`    Response Body
+-- @param `rule_values`   Governance Rule values
+function transformValues(body_table, rule_values)
+    if type(body_table) == "string" then
+        return body_table:gsub("{{%d+}}", rule_values)
+    elseif type(body_table) == "table" and next(body_table) ~= nil then 
+        local updated_body_table = {}
```

Comment:
we had lot of trouble having table inside a while loop.. this looks like recursive function.. I am worried of a leak..if possible avoid it?